### PR TITLE
Initial documentation

### PR DESCRIPTION
**Pull request by kse-clearhaus**

### CODE REVIEW

**Comment by ct-clearhaus** [comment]
Have we thought about leaving the client to decide the `threeDSServerTransID` (or some random-ish thing + UUIDv5) and use this as an "idempotency key"?

File: documentation.md [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+* _Decoupled Authentication_
+
+   This includes 1-4,6-8, where the `authenticationValue`, `eci` and `dsTransID`
+   are returned in 8.
+
+### Performing an authentication flow
+
+1. Use the [`/preauth`](#the-preauth-call) call to determine 3-D Secure version.
+2. Perform an authentication using the [`/auth`](#the-auth-call) call.
```
[the remark this answers is not in the document]
Comment (kse-clearhaus) [reply documentation.md]: That's a fair point. You do mean the endpoints in the API and not in the documentation just, correct?

File: documentation.md [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+TODO: Section is currently, envisioned is a sort of FAQ.
+
+## Reference
+
+### Authentication
+
+Access to the service is granted by an API key. The API key is used on each request using the
+`APIKey` HTTP header as such:
+```
+APIKey: <YOUR UUID API KEY>
```
[the remark this answers is not in the document]
Comment (kse-clearhaus) [reply documentation.md]: The answer is both syntax and semantics.

1. There is no password, there is only the API key. Which is not the same as a user name as a user can have several API keys.
2. I remember several times having to explain to people that they need  to remember the `:` when integrating against the gateway. Though many libraries likely support some HTTP Basic Auth.
3. I don't want someone to perform the call from the browser, e.g. attempt to open the URL and add the API Keys themselves.

But I'm not married to an `APIKey` header.

I found out yesterday that it also matches e.g. what Adyen [does](https://docs.adyen.com/api-explorer/#/PaymentSetupAndVerificationService/v50/overview).

File: documentation.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (ct-clearhaus) [reply documentation.md]: Thinking in terms of HATEOS, these two (the name, and the URL) are two different things, and then I'll care about the name (not the URL) because that is the thing we refer to and document. The URL is "just" the discoverable thing you use as a parameter (together with other parameters like the headers and the body) for your next request 🙂

[the remark this answers is not in the document]
Comment (kse-clearhaus) [reply documentation.md]: You're going to have to dumb it down for me :)

File: documentation.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (ct-clearhaus) [reply documentation.md]: This is an entrance to the JWT consideration (where other systems create long lived JWTs, e.g. with expiry until next payment date).

* Re. 1. and 2. If we have a user concept, then, do we have a username? If so, Basic Auth with `username:apikey` could make sense?

* Re. 3. Why? One could consider it a feature which could be handy for testing — this is basically what's utilized via `https://mpi.3dsecure.io/secure3d_auth`.

**Comment by kse-clearhaus** [comment]
> Have we thought about leaving the client to decide the threeDSServerTransID (or some random-ish thing + UUIDv5) and use this as an "idempotency key"?

Currently the server decides the `threeDSServerTransID`, as a simple way of ensuring a unique ID.

I see the potential for idempotency as rather low, so it was not something I considered.

I feel like there could be some sense in adding idempotency, but that we should wait until we're in proper production mode.
WDYT @clearhaus/pci?

**Comment by ct-clearhaus** [comment]
How does the DS behave if we send the same `threeDSServerTransID` twice? (A reasonable scenario is if the hatred "peer" kills the connection before we received the response.) Is it even spec'ed how the DS should/must behave?

**Comment by kse-clearhaus** [comment]
>How does the DS behave if we send the same threeDSServerTransID twice? (A reasonable scenario is if the hatred "peer" kills the connection before we received the response.) Is it even spec'ed how the DS should/must behave?

They may choose to respond with error code `301`, but there is nothing requiring them to. (Note: Not a HTTP response code, but a specification `Erro` message with code `301`)

File: documentation.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (kse-clearhaus) [reply documentation.md]: * 1. and 2. It could, but we would require the username for no reason, as it is only used in the dashboard. Also: That would mean to semantically do a user lookup by password, which is not the case.
* 3. We can consider creating an endpoint like this, however it is not as simple in 3DSv2. Here we need 2 callback URLs and a lot of data to be entered.

[the remark this answers is not in the document]
Comment (ct-clearhaus) [reply documentation.md]: * Re. 1 and 2: Good point! Also, the API key is not really a password. However, we could use the username as the thing that refers to the user, and then in fact use a password (which we then only store e.g. `scrypt`'ed). However, I like the JWT strategy much better.
* Re. 3: We can just populate it with dummy data (and allow to store data in a cookie). It is quite essential that we can do example flows easily; it'll also be a reference implementation for clients to be inspired by.

File: documentation.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (ct-clearhaus) [reply documentation.md]: Let's IRL it 🙂
